design/cli-install-changes.md

```diff
@@ -0,0 +1,426 @@
+
+To improve the UX for configuring the velero deployment with multiple plugins/providers, and corresponding IAM secrets, the following changes will be made:
+
+- the name of the flag changes from `secret-file` to `velero plugin set --credentials-file`. This is to be consistent with the name of the new flag `velero plugin set --cacert-file`.
+
+- The `velero plugin (create|set) --credentials-file` will be a map of provider name as a key, and the path to the file as a value. This way, we can have multiple credential secrets and each secret per provider/plugin will be unique.
+
+See discussion https://github.com/vmware-tanzu/velero/pull/2259#discussion_r384700723 for the two items below.
+
+- The `velero backup-location (create|set)` will have a new `--credentials mapStringString` flag which sets the name of the corresponding credentials secret for a provider. Format is provider:credentials-secret-name.
+
+- The `velero snapshot-location (create|set)` will have a new `--credentials mapStringString` flag which sets the list of name of the corresponding credentials secret for providers. Format is (provider1:credentials-secret-name1,provider2:credentials-secret-name2,...).
+
+Note that for this logic to work, either the plugin for which the secret corresponds to must have been installed in the system, or we must have a loop checking for when both are present before marking the BSL/VSL as ready.
+
+#### Examples of mounting secrets and environment variables
+
+With the changes proposed in the previous section (Credentials and secrets), the resulting deployment `yaml` would look like below:
+
+AWS
+```
+   spec:
+     containers:
+        volumeMounts:
+        - mountPath: /credentials
+          name: cloud-credentials-aws
+      - args:
+        - server
+        name: velero
+        env:
+        - name: AWS_SHARED_CREDENTIALS_FILE
+          value: /credentials/cloud
+     volumes:
+      - name: cloud-credentials
+        secret:
+          secretName: cloud-credentials
+```
+
+Azure
+```
+   spec:
+     containers:
+        volumeMounts:
+        - mountPath: /credentials
+          name: cloud-credentials-azure
+      - args:
+        - server
+        name: velero
+        env:
+        - name: AZURE_SHARED_CREDENTIALS_FILE
+          value: /credentials/cloud
+     volumes:
+      - name: cloud-credentials-azure
+        secret:
+          secretName: cloud-credentials-azure
+```
+
+
+#### Renaming "provider" to "plugin"
+
+As part of this change, we should change to use the term `plugin` instead of `provider`. The reasoning: in practice, we usually have 1 plugin per provider, and if there is an implementation for both object store and volume snapshotter for that provider, it will all be contained in the same plugin. When we handle plugins, we follow this logic. In other words, there's a plugin name (ex: `velero.io/aws`) and it can contain implementations of kind `ObjectStore` and/or `VolumeSnapshotter`.
```

Comment:
>The reasoning: in practice, we usually have 1 plugin per provider

It depends :)

Some platforms, such as Amazon, Google, and Azure provide object storage and volumes. So they have 2 Velero plugins - one for object storage and one for volume snapshotting ([See the AWS repo as an example](https://github.com/vmware-tanzu/velero-plugin-for-aws/blob/master/velero-plugin-for-aws/main.go#L28-L29), where the 2 plugins are registered inside of 1 executable). vSphere has only a VolumeSnapshotter, though.

These are often shipped together in 1 binary and 1 container image for convenience, but we could also offer separate binaries/images for them.

Additionally, Velero has plugins that aren't related to storage or locations at all - the BackupItemAction and RestoreItemAction plugins are related to the type of Kubernetes object and the action Velero is performing at that time. Thus, using "plugin" could mean one of these types that aren't tied to a location.

Would it perhaps be less ambiguous to say "location plugins"?

Reply:
@nrb I like disambiguating things even further by qualifying it with `location-`. Should it be `location-plugin` or `location-plugins`? I think it should be singular.
WDYT @skriss / @ashish-amarnath?

Reply:
Yeah, I think singular makes sense.